### PR TITLE
Master mdl 71613 59 dev

### PR DESCRIPTION
**Pull request by marcghaly**
Created qbank_viewcreator plugin for issue https://github.com/catalyst/moodle-MDL-70329/issues/59 :

Following was created:

- question/bank/viewcreator/classes/columns.php
- question/bank/viewcreator/classes/creator_name_column.php
- question/bank/viewcreator/classes/modifier_name_column.php
- question/bank/viewcreator/classes/output/renderer.php
- question/bank/viewcreator/classes/privacy/provider.php
- question/bank/viewcreator/lang/en/qbank_viewcreator.php
- question/bank/viewcreator/templates/creator_display.mustache
- question/bank/viewcreator/templates/modifier_display.mustache
- question/bank/viewcreator/version.php

Following was modified: (deprec.)

- question/classes/bank/creator_name_column.php
- question/classes/bank/modifier_name_column.php

### CODE REVIEW

File: question/bank/viewcreator/lang/en/qbank_viewcreator.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * Strings for component qbank_viewcreator, language 'en'
+ *
+ * @package    qbank_viewcreator
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+$string['pluginname'] = 'viewcreator';
```
Comment (safatshahin) [inline question/bank/viewcreator/lang/en/qbank_viewcreator.php:26]: Please use a readable name of the plugin, like View creator

File: question/bank/viewcreator/lang/en/qbank_viewcreator.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+$string['privacy:metadata'] = 'View creator shows the questioncreator column, it does not store any user data.';
```
Comment (safatshahin) [inline question/bank/viewcreator/lang/en/qbank_viewcreator.php:27]: it shows two columns, either mention both or just inform it does not store any user data

File: question/bank/viewcreator/templates/creator_display.mustache [hunk 3]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (safatshahin) [inline question/bank/viewcreator/templates/creator_display.mustache:1]: please remove extra space

File: question/bank/viewcreator/templates/creator_display.mustache [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+{{!
+    @template qbank_viewcreator/creator_display
+    Example context (json):
+    {
+        "displaydata": [
+          {
+                "haslabel": false,
+                "nameclasses": "class-name class-name",
+                "creator": "Creator name",
+                "date": "Date created",
```
Comment (safatshahin) [inline question/bank/viewcreator/templates/creator_display.mustache:24]: please use a real data in the date, example doesnt match

File: question/classes/bank/creator_name_column.php [hunk 5]
```diff
@@ -17,9 +17,11 @@
 /**
  * A column type for the name of the question creator.
  *
- * @package   core_question
- * @copyright 2009 Tim Hunt
- * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @package    core_question
+ * @copyright  2009 Tim Hunt
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0
```
Comment (safatshahin) [inline question/classes/bank/creator_name_column.php:23]: deprecation info will go to the class document

File: question/classes/bank/modifier_name_column.php [hunk 6]
```diff
@@ -20,6 +20,8 @@
  * @package   core_question
  * @copyright 2009 Tim Hunt
  * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0
```
Comment (safatshahin) [inline question/classes/bank/modifier_name_column.php:23]: deprecation info will go to the class document

File: question/bank/viewcreator/templates/creator_display.mustache [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+          }
+        ]
+    }
+}}
+
+{{^haslabel}}
+        {{creator}}
```
Comment (safatshahin) [inline question/bank/viewcreator/templates/creator_display.mustache:31]: the indentation doesn't seem right, can you please double check with your IDE and also it should have 4 spaces, not 8

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+                "modifier": "Modifier name",
+                "date": "Date created",
+          }
+        ]
+    }
+}}
+
+{{^haslabel}}
+        {{modifier}}
+        <br>
```
Comment (safatshahin) [inline question/bank/viewcreator/templates/modifier_display.mustache:32]: please check the indentation, it should be 4

File: question/bank/viewcreator/templates/creator_display.mustache (hunk 7 above)
Comment (safatshahin) [inline question/bank/viewcreator/templates/creator_display.mustache:31]: i would recommend wrap the {{creator}} with a span

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                "nameclasses": "class-name class-name",
+                "modifier": "Modifier name",
+                "date": "Date created",
+          }
+        ]
+    }
+}}
+
+{{^haslabel}}
+        {{modifier}}
```
Comment (safatshahin) [inline question/bank/viewcreator/templates/modifier_display.mustache:31]: i would recommend wrap the {{modifier}} with a span

**Review by safatshahin — COMMENTED** [review]
Hi @marcghaly 
Thank you for working on this, I have added some comments in the code. 
An important note: please add the plugin to core.
(lib/classes/plugin_manager.php)


Regarding the branching, can you please do the following:

Create a branch ahead of master_MDL-71516 naming master_MDL-71613
Then rename the branch master_MDL-71613-59-dev to master_MDL-71613-59

Do a PR from master_MDL-71613-59 to master_MDL-71613

That's how we are aligned with the moodle tracker.

Note: you can do this after the changes and review is done.

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * A column type for the name of the question creator.
+ *
+ * @package    qbank_viewcreator
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:22]: Original copyright used, author added

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * A column type for the name of the question creator.
+ *
+ * @copyright 2009 Tim Hunt
+ * @author    2021 Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class creator_name_column extends column_base {
+    public function get_name(): string {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:40]: Document 6 methods in the class `creator_name_column`

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 12]
```diff
@@ -0,0 +1,34 @@
+
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:1]: Extra space removed

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * A column type for the name of the question last modifier.
+ *
+ * @package    qbank_viewcreator
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:21]: Original copyright used, author added

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    protected function display_content($question, $rowclasses): void {
+        global $OUTPUT, $PAGE;
+        $displaydata = array();
+        $displaydata['haslabel'] = false;
+
+        if (!empty($question->modifierfirstname) && !empty($question->modifierlastname)) {
+            $u = new \stdClass();
+            $u = username_load_fields_from_object($u, $question, 'modifier');
+            $displaydata['date'] = userdate($question->timemodified, get_string('strftimedatetime', 'langconfig'));
+            $displaydata['nameclasses'] = 'date';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:56]: Removed array key value `nameclasses` and added class directly in the mustache file

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    protected function display_content($question, $rowclasses): void {
+        global $OUTPUT, $PAGE;
+        $displaydata = array();
+        $displaydata['haslabel'] = false;
+
+        if (!empty($question->creatorfirstname) && !empty($question->creatorlastname)) {
+            $u = new \stdClass();
+            $u = username_load_fields_from_object($u, $question, 'creator');
+            $displaydata['date'] = userdate($question->timecreated, get_string('strftimedatetime', 'langconfig'));
+            $displaydata['nameclasses'] = 'date';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:57]: Removed array key value `nameclasses` and added class directly in the mustache file

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    public function get_name(): string {
+        return 'creatorname';
+    }
+
+    protected function get_title(): string {
+        return get_string('createdby', 'question');
+    }
+
+    protected function display_content($question, $rowclasses): void {
+        global $OUTPUT, $PAGE;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:49]: $OUTPUT removed

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    public function get_name(): string {
+        return 'modifiername';
+    }
+
+    protected function get_title(): string {
+        return get_string('lastmodifiedby', 'question');
+    }
+
+    protected function display_content($question, $rowclasses): void {
+        global $OUTPUT, $PAGE;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:48]: $OUTPUT removed

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * A column type for the name of the question last modifier.
+ *
+ * @copyright 2009 Tim Hunt
+ * @author    2021 Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class modifier_name_column extends column_base {
+    public function get_name(): string {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:39]: Documented 6 methods in `modifier_name_column` class

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        $displaydata = array();
+        $displaydata['haslabel'] = false;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:51]: Removed key `haslabel`, wasn't sure if that was used in another part of the code

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    protected function get_title(): string {
+        return get_string('lastmodifiedby', 'question');
+    }
+
+    protected function display_content($question, $rowclasses): void {
+        global $OUTPUT, $PAGE;
+        $displaydata = array();
+        $displaydata['haslabel'] = false;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:50]: Removed key `haslabel`

File: question/bank/viewcreator/classes/modifier_name_column.php [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+        if (!empty($question->modifierfirstname) && !empty($question->modifierlastname)) {
+            $u = new \stdClass();
+            $u = username_load_fields_from_object($u, $question, 'modifier');
+            $displaydata['date'] = userdate($question->timemodified, get_string('strftimedatetime', 'langconfig'));
+            $displaydata['nameclasses'] = 'date';
+            $displaydata['modifier'] = fullname($u);
+        }
+        echo $PAGE->get_renderer('qbank_viewcreator')->render_modifier_name($displaydata);
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/modifier_name_column.php:59]: Moved renderer inside if

File: question/bank/viewcreator/classes/creator_name_column.php [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+
+        if (!empty($question->creatorfirstname) && !empty($question->creatorlastname)) {
+            $u = new \stdClass();
+            $u = username_load_fields_from_object($u, $question, 'creator');
+            $displaydata['date'] = userdate($question->timecreated, get_string('strftimedatetime', 'langconfig'));
+            $displaydata['nameclasses'] = 'date';
+            $displaydata['creator'] = fullname($u);
+        }
+        echo $PAGE->get_renderer('qbank_viewcreator')->render_creator_name($displaydata);
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/classes/creator_name_column.php:60]: Moved renderer inside if

File: question/bank/viewcreator/templates/creator_display.mustache [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+                "haslabel": false,
+                "nameclasses": "class-name class-name",
+                "creator": "Creator name",
+                "date": "Date created",
+          }
+        ]
+    }
+}}
+
+{{^haslabel}}
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:30]: Removed useless logic

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+                "haslabel": false,
+                "nameclasses": "class-name class-name",
+                "modifier": "Modifier name",
+                "date": "Date created",
+          }
+        ]
+    }
+}}
+
+{{^haslabel}}
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:30]: Removed useless logic

File: question/bank/viewcreator/templates/creator_display.mustache [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+        {{creator}}
+        <br>
+        <span class="{{nameclasses}}">{{date}}</span>
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:33]: Class `date` used directly

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+        {{modifier}}
+        <br>
+        <span class="{{nameclasses}}">{{date}}</span>
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:33]: Class `date` used directly

File: question/bank/viewcreator/templates/modifier_display.mustache [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+{{!
+    @template qbank_viewcreator/modifier_display
+    Example context (json):
+    {
+        "displaydata": [
+          {
+                "haslabel": false,
+                "nameclasses": "class-name class-name",
+                "modifier": "Modifier name",
+                "date": "Date created",
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:24]: Example date and modifier changed to real ones

File: question/bank/viewcreator/lang/en/qbank_viewcreator.php (hunk 1 above)
Comment (marcghaly) [reply question/bank/viewcreator/lang/en/qbank_viewcreator.php:26]: Used readable name of the plugin

File: question/bank/viewcreator/lang/en/qbank_viewcreator.php (hunk 2 above)
Comment (marcghaly) [reply question/bank/viewcreator/lang/en/qbank_viewcreator.php:27]: Informed that it shows creator and modifier column

File: question/bank/viewcreator/templates/creator_display.mustache (hunk 3 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:1]: Extra space removed

File: question/bank/viewcreator/templates/creator_display.mustache (hunk 4 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:24]: Real data used for examples and removed useless logic

File: question/bank/viewcreator/templates/creator_display.mustache (hunk 7 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:31]: Fixed indentation, checked IDE tabsize is set to 4, my bad

File: question/bank/viewcreator/templates/modifier_display.mustache (hunk 8 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:32]: Fixed indentation

File: question/bank/viewcreator/templates/creator_display.mustache (hunk 7 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/creator_display.mustache:31]: Wrapped `{{creator}}` in a span

File: question/bank/viewcreator/templates/modifier_display.mustache (hunk 9 above)
Comment (marcghaly) [reply question/bank/viewcreator/templates/modifier_display.mustache:31]: Wrapped `{{modifier}}` in a span

File: question/classes/bank/creator_name_column.php (hunk 5 above)
Comment (marcghaly) [reply question/classes/bank/creator_name_column.php:23]: Moved depreciation in class documentation

File: question/classes/bank/modifier_name_column.php (hunk 6 above)
Comment (marcghaly) [reply question/classes/bank/modifier_name_column.php:23]: Moved depreciation in class documentation

**Comment by marcghaly** [comment]
Conversation resolution, modified following:

- question/bank/viewcreator/classes/creator_name_column.php
- question/bank/viewcreator/classes/modifier_name_column.php
- question/bank/viewcreator/lang/en/qbank_viewcreator.php
- question/bank/viewcreator/templates/creator_display.mustache
- question/bank/viewcreator/templates/modifier_display.mustache
- question/classes/bank/creator_name_column.php
- question/classes/bank/modifier_name_column.php

**Pull request merged**